fs/fs.go

```diff
@@ -74,20 +74,81 @@ import (
 	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/semaphore"
 )
 
 var (
 	defaultIndexSelectionPolicy = SelectFirstPolicy
 	fusermountBin               = "fusermount"
+	preresolverQueueBufferSize  = 1024 // arbitrarily chosen buffer size
 )
 
+// Preresolver will resolve a number of layers in parallel,
+// up to the amount specified by MaxConcurrency.
+type preresolver struct {
+	queue chan func(context.Context) string
+	cache *sync.Map
+	smp   *semaphore.Weighted
+}
+
+func newPreresolver(maxConcurrency int64) *preresolver {
+	pr := &preresolver{}
+	pr.queue = make(chan func(context.Context) string, preresolverQueueBufferSize)
+	pr.cache = &sync.Map{}
+	if maxConcurrency > 0 {
+		pr.smp = semaphore.NewWeighted(maxConcurrency)
+	}
+	return pr
+}
+
+func (pr *preresolver) Start(ctx context.Context) error {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				resolveFn := <-pr.queue
+				// If concurrency limits are disabled,
+				// we don't need to wait for a semaphore
+				if pr.smp != nil {
```

Comment:
I wonder if instead of checking if the semaphore is nil here, we only create a `preresolver` if the max concurrency is > 0 in fs.go. We can check if `preresolver != nil` before getting into the pre-resolve codeblock [here](https://github.com/awslabs/soci-snapshotter/pull/1061/files#diff-c7b54bb28e1654d0ca47364a2c97456c124ee0d3a65854db18d116c509d89136L444). This way we don't have an unnecessary struct and go routine in memory.

Reply:
The preresolver still handles deduplication in [Enqueue()](https://github.com/awslabs/soci-snapshotter/blob/8b303bdb3b1d214d2671f2b43320a616f5960ee3/fs/fs.go#L132), so I think we should keep this even with concurrency limits disabled.